File: src/server/mcpClients.ts
import { randomUUID } from "node:crypto";
import type { Implementation } from "@modelcontextprotocol/sdk/types.js";
import type { SceneSelection } from "../shared/schema";
import type { WorldRect } from "../shared/renderScene";
import type { ExtendedRenderPatch } from "../shared/renderPatch";

// ---------------------------------------------------------------------------
// T5.1: Client identity and dock state
// ---------------------------------------------------------------------------

export type McpClientDockState = "idle" | "active" | "error" | "disconnected" | "muted";

// T5.2: Canvas target type (widened from T5.1's SceneSelection | {kind:"viewport"} | null)
export type CanvasTarget =
  | { kind: "canvas" }
  | { kind: "group"; id: string }
  | { kind: "node"; id: string }
  | { kind: "edge"; id: string }
  | { kind: "selection"; ids: SceneSelection[] }
  | { kind: "viewport"; rect: WorldRect }
  | { kind: "artifact"; artifactId: string; groupId: string };

export type McpClientIdentity = {
  clientId: string;
  actorType: "mcp";
  label: string;
  name: string;
  version: string;
  color: string;
  iconRef: string | null;
  transport: "stdio" | "http";
  dockState: McpClientDockState;
  lastTarget: CanvasTarget | null;
  connectedAt: number;
  lastActivityAt: number;
  muted: boolean;
  /** T5.4: bounded ephemeral ring of recent read + error events (never persisted). */
  readRing: ReadTraceEvent[];
};

// ---------------------------------------------------------------------------
// T5.4: Trace model — types
// ---------------------------------------------------------------------------

/**
 * Every trace category named in the T5.4 design fragment.
 * Write/comment/export/proposal-* are projected from the events table on
 * demand; only "read" and "error" require the in-memory readRing.
 */
export type TraceKind =
  | "read"
  | "write"
  | "comment"
  | "export"
  | "proposal-created"
  | "proposal-accepted"
  | "proposal-rejected"
  | "error";

/** A single projected trace event (display-only; never persisted). */
export type TraceEvent = {
  clientId: string;
  kind: TraceKind;
  target: CanvasTarget;
  verb: string;
  /** events.id for logged ops; null for reads / pre-commit errors. */
  operationId: string | null;
  /** OperationEnvelope.sourceToolCall.callId — links trace back to the MCP call. */
  callId: string | null;
  /** Epoch ms. */
  at: number;
  /** Short human-readable summary, recomputed on read, never persisted. */
  summary: string;
  errorMessage: string | null;
};

/**
 * One entry in the per-client read-event ring.
 * Written by the tracking middleware for query_scene / list_groups / get_group.
 */
export type ReadTraceEvent = {
  tool: string;
  target: CanvasTarget;
  at: number;
  errorMessage?: string;
};

/** Maximum number of read/error events kept per client ring. */
export const READ_RING_CAPACITY = 50;

// ---------------------------------------------------------------------------
// T5.4: Ghost write-preview shape
// ---------------------------------------------------------------------------

/**
 * Represents a pending risky-write preview (ghost diff).
 *
 * Built server-side by applying the staged OperationEnvelope.patch to a
 * throwaway copy of the scene via applyRenderPatchToShapeScene (pure, no
 * commit).  Carried as metadata alongside a pending proposal; the actual
 * geometry rendering is deferred to the shell/core overlay layer (§3 of the
 * design fragment).
 */
export type WritePreview = {
  /** The pending proposal this preview corresponds to. */
  proposalId: string;
  clientId: string;
  /** The risk classification that routed this write to staging. */
  riskClass: "destructive" | "wide" | "agent-flagged";
  /** The patch verb (kind) that would be committed on accept. */
  verb: string;
  /** Semantic target ids derived from the staged patch. */
  targetIds: string[];
  /** Epoch ms when the proposal was staged. */
  stagedAt: number;
};

// ---------------------------------------------------------------------------
// Color palette — 12 distinct hues, deterministic per clientId
// ---------------------------------------------------------------------------

const COLOR_PALETTE = [
  "#e05252", "#e0893d", "#d4be34", "#52b765",
  "#3db8b8", "#3d82e0", "#7b52e0", "#c252c2",
  "#e07b7b", "#7be07b", "#7bc4e0", "#c2a852"
];

function hashClientId(clientId: string): number {
  let h = 0;
  for (let i = 0; i < clientId.length; i++) {
    h = (Math.imul(31, h) + clientId.charCodeAt(i)) >>> 0;
  }
  return h;
}

function colorFromClientId(clientId: string): string {
  return COLOR_PALETTE[hashClientId(clientId) % COLOR_PALETTE.length];
}

// ---------------------------------------------------------------------------
// Registry
// ---------------------------------------------------------------------------

const _clients = new Map<string, McpClientIdentity>();

export function getClientRegistry(): ReadonlyMap<string, McpClientIdentity> {
  return _clients;
}

export function listClients(): McpClientIdentity[] {
  return Array.from(_clients.values());
}

export function getClient(clientId: string): McpClientIdentity | undefined {
  return _clients.get(clientId);
}

/**
 * Create and register a new McpClientIdentity from an SDK Implementation.
 * Returns the new identity record.
 */
export function registerClient(
  impl: Implementation,
  transport: "stdio" | "http",
  clientId?: string
): McpClientIdentity {
  const id = clientId ?? randomUUID();
  const rawLabel = impl.title ?? impl.name;
  const label = deduplicateLabel(rawLabel, id);
  const now = Date.now();
  const identity: McpClientIdentity = {
    clientId: id,
    actorType: "mcp",
    label,
    name: impl.name,
    version: impl.version,
    color: colorFromClientId(id),
    iconRef: impl.icons?.[0]?.src ?? null,
    transport,
    dockState: "idle",
    lastTarget: null,
    connectedAt: now,
    lastActivityAt: now,
    muted: false,
    readRing: []
  };
  _clients.set(id, identity);
  return identity;
}

/** Mark a client's dock state. */
export function setClientDockState(
  clientId: string,
  state: McpClientDockState,
  errorMessage?: string
): void {
  const client = _clients.get(clientId);
  if (!client) return;
  client.dockState = state;
  client.lastActivityAt = Date.now();
  // errorMessage is informational; stored for future inspection (e.g. tooltip)
  void errorMessage;
}

/** Update the lastTarget for a client. */
export function setClientLastTarget(clientId: string, target: CanvasTarget | null): void {
  const client = _clients.get(clientId);
  if (!client) return;
  client.lastTarget = target;
}

/** Remove a client from the registry (called after disconnect grace). */
export function removeClient(clientId: string): void {
  _clients.delete(clientId);
}

/** For testing: reset the registry. */
export function _resetClientRegistry(): void {
  _clients.clear();
}

// ---------------------------------------------------------------------------
// T5.4: Read-ring mutation helpers
// ---------------------------------------------------------------------------

/**
 * Push a read trace event into the per-client ring (for query_scene /
 * list_groups / get_group).  Oldest entry is evicted when the ring is full.
 */
export function pushReadTrace(clientId: string, event: ReadTraceEvent): void {
  const client = _clients.get(clientId);
  if (!client) return;
  client.readRing.push(event);
  if (client.readRing.length > READ_RING_CAPACITY) {
    client.readRing.shift();
  }
}

/**
 * Push an error trace event into the per-client ring.
 * Uses tool name as `tool` and `errorMessage` for display.
 */
export function pushErrorTrace(
  clientId: string,
  tool: string,
  target: CanvasTarget,
  errorMessage: string
): void {
  pushReadTrace(clientId, {
    tool,
    target,
    at: Date.now(),
    errorMessage
  });
}

// ---------------------------------------------------------------------------
// T5.4: Risk classifier — routes risky writes to staged proposals
// ---------------------------------------------------------------------------

const DESTRUCTIVE_KINDS = new Set<string>([
  "delete-group",
  "delete-card",
  "delete-edge"
]);

/** Wide-write threshold: patch touching this many targets or more is "wide". */
const WIDE_TARGET_THRESHOLD = 5;

/**
 * Classify a patch as risky (destructive or wide) or safe (immediate commit).
 *
 * Returns the risk class if the write should be staged as a proposal, or null
 * if the write should be committed immediately.
 */
export function classifyWriteRisk(
  patch: ExtendedRenderPatch,
  targetIds: string[]
): WritePreview["riskClass"] | null {
  if (DESTRUCTIVE_KINDS.has(patch.kind)) return "destructive";
  if (targetIds.length >= WIDE_TARGET_THRESHOLD) return "wide";
  return null;
}

// ---------------------------------------------------------------------------
// T5.4: Trace query — project events log + read ring per client
// ---------------------------------------------------------------------------

/**
 * Derive a human-readable summary for a trace event.
 * Display-only; recomputed on read, never persisted.
 */
export function deriveTraceSummary(kind: TraceKind, verb: string, target: CanvasTarget): string {
  const targetDesc = targetDescription(target);
  switch (kind) {
    case "read":
      return targetDesc ? `read ${targetDesc}` : `read ${verb}`;
    case "write":
      return targetDesc ? `${verb} ${targetDesc}` : verb;
    case "comment":
      return targetDesc ? `commented on ${targetDesc}` : "commented";
    case "export":
      return targetDesc ? `exported ${targetDesc}` : "exported";
    case "proposal-created":
      return targetDesc ? `proposed: ${verb} ${targetDesc} (pending)` : `proposed: ${verb} (pending)`;
    case "proposal-accepted":
      return targetDesc ? `accepted proposal → ${verb} ${targetDesc}` : `accepted proposal → ${verb}`;
    case "proposal-rejected":
      return "rejected proposal (no change)";
    case "error":
      return `error on ${targetDesc ?? verb}`;
  }
}

function targetDescription(target: CanvasTarget): string | null {
  switch (target.kind) {
    case "group":
      return `group ${target.id}`;
    case "node":
      return `node ${target.id}`;
    case "edge":
      return `edge ${target.id}`;
    case "artifact":
      return `artifact ${target.artifactId}`;
    case "selection":
      return `selection (${target.ids.length} items)`;
    case "viewport":
      return "viewport";
    case "canvas":
      return null;
  }
}

/**
 * Project the per-client read ring into TraceEvent records.
 * Read ring entries with errorMessage become kind:"error"; others are kind:"read".
 */
export function projectReadRing(clientId: string): TraceEvent[] {
  const client = _clients.get(clientId);
  if (!client) return [];
  return client.readRing.map((entry) => {
    const kind: TraceKind = entry.errorMessage ? "error" : "read";
    return {
      clientId,
      kind,
      target: entry.target,
      verb: entry.tool,
      operationId: null,
      callId: null,
      at: entry.at,
      summary: entry.errorMessage
        ? `error: ${entry.errorMessage}`
        : deriveTraceSummary("read", entry.tool, entry.target),
      errorMessage: entry.errorMessage ?? null
    };
  });
}

/**
 * Project one raw events-table row into a TraceEvent.
 * `row` is the deserialized payload_json (OperationEnvelope shape).
 */
export function projectEventRow(clientId: string, row: {
  id: string;
  type: string;
  payloadJson: string;
  createdAt: string;
}): TraceEvent {
  let payload: Record<string, unknown> = {};
  try {
    payload = JSON.parse(row.payloadJson) as Record<string, unknown>;
  } catch {
    // malformed payload — fall through with defaults
  }

  const kind = eventTypeToTraceKind(row.type);
  const targetIds = Array.isArray(payload.targetIds) ? (payload.targetIds as string[]) : [];
  const target = targetIdsToCanvasTarget(targetIds);
  const callId = (payload.sourceToolCall as { callId?: string } | undefined)?.callId ?? null;
  const at = new Date(row.createdAt).getTime();

  return {
    clientId,
    kind,
    target,
    verb: row.type,
    operationId: row.id,
    callId,
    at,
    summary: deriveTraceSummary(kind, row.type, target),
    errorMessage: null
  };
}

function eventTypeToTraceKind(type: string): TraceKind {
  if (type === "add-comment") return "comment";
  if (type === "export") return "export";
  if (type === "accept-proposal") return "proposal-accepted";
  if (type === "reject-proposal") return "proposal-rejected";
  // proposal-created is only surfaced via the staged sceneProposalSchema row,
  // not via an events row, so it won't appear here.
  return "write";
}

function targetIdsToCanvasTarget(targetIds: string[]): CanvasTarget {
  if (targetIds.length === 0) return { kind: "canvas" };
  if (targetIds.length === 1) {
    const id = targetIds[0];
    // Heuristic: prefix-based kind detection matching scene id conventions.
    if (id.startsWith("g")) return { kind: "group", id };
    if (id.startsWith("e")) return { kind: "edge", id };
    return { kind: "node", id };
  }
  // Multiple targets: represent as a selection of node-kind (display heuristic).
  return {
    kind: "selection",
    ids: targetIds.map((id) => {
      if (id.startsWith("g")) return { kind: "group" as const, id };
      if (id.startsWith("e")) return { kind: "edge" as const, id };
      return { kind: "node" as const, id };
    })
  };
}

// ---------------------------------------------------------------------------
// Label collision resolution: append " (2)", " (3)" etc. by connect order
// ---------------------------------------------------------------------------

function deduplicateLabel(base: string, newId: string): string {
  const existing = Array.from(_clients.values()).filter(
    (c) => c.clientId !== newId && (c.label === base || c.label.startsWith(base + " ("))
  );
  if (existing.length === 0) return base;
  return `${base} (${existing.length + 1})`;
}

// ---------------------------------------------------------------------------
// T5.2: Tool → CanvasTarget mapping
// ---------------------------------------------------------------------------

export type ToolName =
  | "query_scene"
  | "list_groups"
  | "get_group"
  | "create_group"
  | "patch_scene"
  | "create_tag"
  | "update_group_tags"
  | "set_selection"
  | "add_comment"
  | "export_group";

export type ToolCallInfo = {
  tool: ToolName;
  /** groupId arg (get_group, update_group_tags, export_group) */
  groupId?: string;
  /** SceneSelection arg (set_selection, add_comment target) */
  selection?: SceneSelection;
  /** tag filter (query_scene) */
  tagIds?: string[];
  /** ids from T2.5 envelope targetIds (patch_scene, create_group result, etc.) */
  targetIds?: SceneSelection[];
  /** artifact result (export_group) */
  artifactId?: string;
};

/**
 * Resolve a tool call to a CanvasTarget.
 *
 * Resolution order (first match wins):
 *   1. envelope targetIds present → spatial CanvasTarget
 *   2. read tool with group arg → group/viewport
 *   3. whole-scene read → scene-fit viewport (represented as {kind:"canvas"})
 *   4. registry/non-spatial write → {kind:"canvas"}
 *
 * Falls back to null when no target can be determined (keeps prior lastTarget).
 */
export function resolveToolTarget(info: ToolCallInfo): CanvasTarget | null {
  const { tool, groupId, selection, targetIds, artifactId } = info;

  switch (tool) {
    case "query_scene":
      // viewport: hull of matched groups (approximated as canvas-wide for now)
      // A full hull computation requires scene data; caller may pass a rect via
      // targetIds[0] of kind "group" for a filtered read. If none, whole-scene.
      return { kind: "canvas" };

    case "list_groups":
      // whole-scene survey
      return { kind: "canvas" };

    case "get_group":
      if (groupId) return { kind: "group", id: groupId };
      return null;

    case "create_group":
      if (targetIds && targetIds.length === 1) return selectionToTarget(targetIds[0]);
      return { kind: "canvas" };

    case "patch_scene": {
      if (!targetIds || targetIds.length === 0) return null;
      if (targetIds.length === 1) return selectionToTarget(targetIds[0]);
      return { kind: "selection", ids: targetIds };
    }

    case "create_tag":
      // registry/non-spatial write — no canvas location
      return { kind: "canvas" };

    case "update_group_tags":
      if (groupId) return { kind: "group", id: groupId };
      return { kind: "canvas" };

    case "set_selection":
      if (selection) return selectionToTarget(selection);
      return null;

    case "add_comment":
      if (selection) return selectionToTarget(selection);
      return { kind: "canvas" };

    case "export_group":
      if (artifactId && groupId) return { kind: "artifact", artifactId, groupId };
      if (groupId) return { kind: "group", id: groupId };
      return null;

    default:
      return null;
  }
}

function selectionToTarget(sel: SceneSelection): CanvasTarget {
  return sel as CanvasTarget;
}

File: crates/server/src/mcp_clients.rs
//! MG2.3: in-memory MCP client registry + per-client trace ring.
//!
//! Port of the dock-facing half of the Node `src/server/mcpClients.ts`: a live
//! map of connected MCP companions (one entry per session), a stable per-client
//! colour derived from a FNV-like hash, and a bounded ring of recent read/write
//! trace events the companion dock polls over HTTP. This is display-only state
//! that is never persisted — it disappears with the process.

use std::collections::HashMap;
use std::sync::{Arc, Mutex};
use std::time::{SystemTime, UNIX_EPOCH};

use serde::Serialize;

/// 12 distinct hues, deterministic per clientId — identical palette to the Node
/// dock so a companion keeps its colour across the Node→Rust cutover.
const COLOR_PALETTE: [&str; 12] = [
    "#e05252", "#e0893d", "#d4be34", "#52b765", "#3db8b8", "#3d82e0", "#7b52e0", "#c252c2",
    "#e07b7b", "#7be07b", "#7bc4e0", "#c2a852",
];

/// FNV-like 32-bit hash matching the Node `hashClientId` (`h = imul(31,h)+c`).
fn hash_client_id(client_id: &str) -> u32 {
    let mut h: u32 = 0;
    for byte in client_id.bytes() {
        h = 31u32.wrapping_mul(h).wrapping_add(byte as u32);
    }
    h
}

/// Stable colour for a client id from the 12-hue palette.
pub fn color_from_client_id(client_id: &str) -> &'static str {
    COLOR_PALETTE[(hash_client_id(client_id) as usize) % COLOR_PALETTE.len()]
}

fn now_ms() -> u64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_millis() as u64)
        .unwrap_or(0)
}

/// One connected MCP companion, as the dock sees it.
#[derive(Clone, Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct McpClientIdentity {
    pub client_id: String,
    pub actor_type: &'static str,
    pub label: String,
    pub name: String,
    pub version: String,
    pub color: &'static str,
    pub transport: &'static str,
    pub dock_state: &'static str,
    pub connected_at: u64,
    pub last_activity_at: u64,
}

/// Kind of a projected trace event (a subset of the Node `TraceKind`; the Rust
/// server tracks read/write/comment/export/error).
#[derive(Clone, Copy, Debug, Serialize)]
#[serde(rename_all = "kebab-case")]
pub enum TraceKind {
    Read,
    Write,
    Comment,
    Export,
    Error,
}

/// One display-only trace event. Recomputed on read; never persisted.
#[derive(Clone, Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct TraceEvent {
    pub client_id: String,
    pub kind: TraceKind,
    pub verb: String,
    pub at: u64,
    pub summary: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error_message: Option<String>,
}

/// Maximum number of trace events kept per client ring (matches the Node
/// `READ_RING_CAPACITY`).
const RING_CAPACITY: usize = 50;

struct ClientEntry {
    identity: McpClientIdentity,
    ring: Vec<TraceEvent>,
}

/// Shared, clone-able registry of connected MCP companions + their trace rings.
#[derive(Clone, Default)]
pub struct ClientRegistry {
    inner: Arc<Mutex<HashMap<String, ClientEntry>>>,
}

impl ClientRegistry {
    pub fn new() -> Self {
        Self::default()
    }

    /// Register (or refresh) a companion from its advertised name/version.
    /// `label` de-duplicates by connect order, mirroring the Node registry.
    pub fn register(&self, client_id: &str, name: &str, version: &str, transport: &'static str) {
        let mut map = self.inner.lock().expect("client registry mutex poisoned");
        let label = deduplicate_label(&map, name, client_id);
        let now = now_ms();
        let identity = McpClientIdentity {
            client_id: client_id.to_string(),
            actor_type: "mcp",
            label,
            name: name.to_string(),
            version: version.to_string(),
            color: color_from_client_id(client_id),
            transport,
            dock_state: "idle",
            connected_at: now,
            last_activity_at: now,
        };
        map.entry(client_id.to_string())
            .and_modify(|e| {
                e.identity.name = name.to_string();
                e.identity.version = version.to_string();
                e.identity.last_activity_at = now;
                e.identity.dock_state = "idle";
            })
            .or_insert(ClientEntry {
                identity,
                ring: Vec::new(),
            });
    }

    /// Snapshot every connected companion (dock list).
    pub fn list(&self) -> Vec<McpClientIdentity> {
        self.inner
            .lock()
            .expect("client registry mutex poisoned")
            .values()
            .map(|e| e.identity.clone())
            .collect()
    }

    /// Push one trace event into a client's ring, evicting the oldest when full,
    /// and bump its last-activity / dock state.
    pub fn push_trace(
        &self,
        client_id: &str,
        kind: TraceKind,
        verb: &str,
        summary: String,
        error_message: Option<String>,
    ) {
        let mut map = self.inner.lock().expect("client registry mutex poisoned");
        let Some(entry) = map.get_mut(client_id) else {
            return;
        };
        let at = now_ms();
        entry.identity.last_activity_at = at;
        entry.identity.dock_state = if error_message.is_some() {
            "error"
        } else {
            "active"
        };
        entry.ring.push(TraceEvent {
            client_id: client_id.to_string(),
            kind,
            verb: verb.to_string(),
            at,
            summary,
            error_message,
        });
        if entry.ring.len() > RING_CAPACITY {
            entry.ring.remove(0);
        }
    }

    /// Recent trace for one client, newest-first, capped at `limit`.
    pub fn trace(&self, client_id: &str, limit: usize) -> Vec<TraceEvent> {
        let map = self.inner.lock().expect("client registry mutex poisoned");
        let Some(entry) = map.get(client_id) else {
            return Vec::new();
        };
        let mut out: Vec<TraceEvent> = entry.ring.iter().rev().cloned().collect();
        out.truncate(limit);
        out
    }

    /// Mark a companion disconnected (kept so a late trace read still resolves).
    pub fn mark_disconnected(&self, client_id: &str) {
        if let Some(entry) = self
            .inner
            .lock()
            .expect("client registry mutex poisoned")
            .get_mut(client_id)
        {
            entry.identity.dock_state = "disconnected";
            entry.identity.last_activity_at = now_ms();
        }
    }
}

/// Append " (2)", " (3)" … when a companion's name collides with a connected
/// one, by connect order — a faithful port of the Node `deduplicateLabel`.
fn deduplicate_label(map: &HashMap<String, ClientEntry>, base: &str, new_id: &str) -> String {
    let prefix = format!("{base} (");
    let existing = map
        .values()
        .filter(|e| {
            e.identity.client_id != new_id
                && (e.identity.label == base || e.identity.label.starts_with(&prefix))
        })
        .count();
    if existing == 0 {
        base.to_string()
    } else {
        format!("{base} ({})", existing + 1)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn color_is_stable_and_from_palette() {
        let c1 = color_from_client_id("abc");
        let c2 = color_from_client_id("abc");
        assert_eq!(c1, c2);
        assert!(COLOR_PALETTE.contains(&c1));
    }

    #[test]
    fn hash_matches_node_fnv_like() {
        // h = imul(31, h) + charCode, unsigned 32-bit.
        // "a" => 97
        assert_eq!(hash_client_id("a"), 97);
        // "ab" => 31*97 + 98 = 3105
        assert_eq!(hash_client_id("ab"), 3105);
    }

    #[test]
    fn register_then_list_and_trace() {
        let reg = ClientRegistry::new();
        reg.register("s1", "claude", "1.0", "http");
        reg.register("s1", "claude", "1.0", "http");
        let list = reg.list();
        assert_eq!(list.len(), 1);
        assert_eq!(list[0].label, "claude");

        reg.push_trace("s1", TraceKind::Write, "create-group", "made g1".into(), None);
        let trace = reg.trace("s1", 50);
        assert_eq!(trace.len(), 1);
        assert_eq!(trace[0].verb, "create-group");
    }

    #[test]
    fn label_dedup_on_collision() {
        let reg = ClientRegistry::new();
        reg.register("s1", "claude", "1", "http");
        reg.register("s2", "claude", "1", "http");
        let labels: Vec<String> = reg.list().into_iter().map(|c| c.label).collect();
        assert!(labels.contains(&"claude".to_string()));
        assert!(labels.contains(&"claude (2)".to_string()));
    }
}
